src/boot/efi/boot.c

```diff
@@ -1643,12 +1652,41 @@ static INTN config_entry_compare(const ConfigEntry *a, const ConfigEntry *b) {
         assert(a);
         assert(b);
 
-        /* Order entries that have no tries left to the beginning of the list */
-        if (a->tries_left != 0 && b->tries_left == 0)
-                return 1;
+        /* Order entries that have no tries left to the end of the list */
         if (a->tries_left == 0 && b->tries_left != 0)
+                return 1;
+        if (a->tries_left != 0 && b->tries_left == 0)
                 return -1;
 
+        /* If there's a sort key defined for *both* entries, then we do new-style ordering, i.e. by
+         * sort-key/machine-id/version, with a final fallback to id. If there's no sort key for either, we do
+         * old-style ordering, i.e. by id only. If one has sort key and the other does not, we put new-style
+         * before old-style. */
+        if (a->sort_key) {
+
+                if (!b->sort_key) /* a is new-style but b old-style */
+                        return -1;
+
+                r = strcmp(a->sort_key, b->sort_key);
+                if (r != 0)
+                        return r;
+
+                /* If multiple installations of the same OS are around, group by machine ID */
+                r = strcmp_ptr(a->machine_id, b->machine_id);
+                if (r != 0)
+                        return r;
+
+                /* If the sort key was defined, then order by version now (downwards, putting the newest first) */
+                r = -strverscmp_improved(a->version, b->version);
+                if (r != 0)
+                        return r;
+
+        } else if (b->sort_key) /* a is old-style but b new-style */
```

Comment:
I don't like how this is expressed, putting this branch so far away. Keeping it similar to how the boot counting comparisons are done would be nicer, no?
```c
if (a && !b)
    return -1;
if (!a && b)
    return 1;
if (a && b)
    .....

Reply:
sure